urllib3/connection.py

```diff
@@ -299,7 +302,18 @@ def connect(self):
                     'for details.)'.format(hostname)),
                     SubjectAltNameWarning
                 )
-            match_hostname(cert, self.assert_hostname or hostname)
+            asserted_hostname = self.assert_hostname or hostname
+            try:
+                match_hostname(cert, asserted_hostname)
+            except CertificateError as e:
+                log.error(
+                    'Certificate did not match expected hostname: %s. '
+                    'Certificate: %s', asserted_hostname, cert
+                )
+                # Add cert to exception and reraise so client code can inspect
+                # the cert when catching the exception, if they want to
+                e.cert = cert
```

Comment:
Hm, maybe we should be centralizing the SSL hostname matching logic elsewhere so we can point people towards the tools to use instead of implementing this kind of logic from a vendored package themselves?

That said, if we are raising this exception to the user (which I'm fairly certain we aren't) we should stop that and replace it with a different exception.